semgrep/semgrep/target_manager.py

```diff
@@ -62,26 +67,22 @@ def lang_to_exts(language: Language) -> List[FileExtension]:
         raise _UnknownLanguageError(f"Unsupported Language: {language}")
 
 
+@attr.s(auto_attribs=True)
 class TargetManager:
-    def __init__(
-        self,
-        includes: List[str],
-        excludes: List[str],
-        targets: List[str],
-        respect_git_ignore: bool,
-    ) -> None:
-        """
-            Handles all file include/exclude logic for semgrep
+    """
+        Handles all file include/exclude logic for semgrep
 
-            If respect_git_ignore is true then will only consider files that are
-            tracked or (untracked but not ignored) by git
-        """
-        self._targets = targets
-        self._includes = includes
-        self._excludes = excludes
-        self._respect_git_ignore = respect_git_ignore
+        If respect_git_ignore is true then will only consider files that are
+        tracked or (untracked but not ignored) by git
+    """
+
+    targets: List[str]
+    includes: List[str]
+    excludes: List[str]
+    respect_git_ignore: bool
+    output_handler: OutputHandler
```

Comment:
TargetManager should not need to know how to handle output. Probably better to have functions return relevant info and the caller can handle how it wants to deal with outputing said info.

Reply:
Can you elaborate on this a bit more? I understand the preference for not making output_handler a required argument, but don't really understand the alternative you are suggesting (how would you change the FilesNotFoundError?)

Reply:
@brendongo IMO we should really have a context object do this (a la Bento) or a reader monad or the like. But passing output_handler here is exactly how we separate knowing how to handle output from the TargetManager.

Reply:
I mean: TargetManager should concern itself with file includes excludes and not need to know how to use an output_handler object. Resolved targets can return a tuple with the second item being problematic files.

Reply:
TargetManager should not need to know that any errors are "handled" or that they are even errors to begin with. The TargetManager just knows that of the files you handed it, these happened to not exist and it lets you know that. It's up to the caller to deal with this information.

Reply:
Thinking about this more and the list of files that don't exist is unrelated to the language so might be better placed in the constructor instead of in this code path. Seems like a significant endeavor would be supportive of getting this PR in as is with injection implementation and revisiting that and constructor in a follow up.

Reply:
you might want to think about the edge cases of a rule with multiple languages calling get_files multiple times though

Reply:
Good thought! We just tested this case and it seems to print out the right thing (only once)